File: src/nunja/engine.py
# -*- coding: utf-8 -*-
import codecs
import json
from functools import partial

from jinja2 import Environment

from calmjs.registry import get
from nunja.registry import REQ_TMPL_NAME
from nunja.registry import DEFAULT_WRAPPER_NAME
from nunja.registry import DEFAULT_REGISTRY_NAME
from nunja.registry import JINJA_TEMPLATE_REGISTRY_NAME
from nunja.registry import MoldRegistry
from nunja.registry import JinjaTemplateRegistry
from nunja.loader import NunjaLoader


def join(*p):
    # the engine internally use standard path separator characters, the
    # registry is responsible for resolving that to the underlying path
    # on the actual filesystem.
    return '/'.join(p)


class Engine(object):
    """
    Nunja core engine

    This takes a nunja mold registry and is able to execute the
    rendering of templates through nunja identifiers.
    """

    def __init__(
            self,
            registry=DEFAULT_REGISTRY_NAME,
            env=None,
            _wrapper_name=DEFAULT_WRAPPER_NAME,
            _required_template_name=REQ_TMPL_NAME,
            ):
        """
        By default, the engine can be created without arguments which
        will initialize using the default registry.

        It is possible to initialize using other arguments, but this is
        unsupported by the main system, and only useful for certain
        specialized implementations.
        """

        self.registry = (
            registry if isinstance(registry, MoldRegistry) else get(registry))
        self.env = env if env else Environment(
            autoescape=True,
            loader=NunjaLoader(self.registry)
        )
        # this filter is to match with nunjucks version (which calls
        # JSON.stringify in JavaScript); construct a partial which is a
        # callable to json.dumps with default parameters that mimic the
        # JavaScript version of the called function.
        self.env.filters['dump'] = partial(
            json.dumps, sort_keys=True, separators=(',', ':'))
        self._required_template_name = _required_template_name

        self._core_template_ = self.load_mold(_wrapper_name)

    def lookup_path(self, name):
        """
        Lookup the path of the underlying resource identified by name
        through the loader.
        """

        return self.registry.verify_path(name)

    def fetch_path(self, name):
        """
        Fetch contents from the path retrieved via lookup_path.

        No caching will be done.
        """

        with codecs.open(self.lookup_path(name), encoding='utf-8') as fd:
            return fd.read()

    def load_template(self, name):
        """
        Load the template identified by name as found in the registry.
        """

        return self.env.get_template(name)

    def load_mold(self, mold_id):
        """
        Load the default, required template from the mold `mold_id`.

        This is useful to explicitly render something without being
        wrapped by the helper div to avoid the automatic client-side
        execution hooks.  Example:

            engine.load_mold('nunja.molds/html5').render(title='Hello')
        """

        return self.load_template(join(mold_id, self._required_template_name))

    def execute(self, mold_id, data, wrapper_tag='div'):
        """
        Execute a mold `mold_id` by rendering through ``env``.

        This is done using its default template, with data provided as
        dict.

        This returns the wrapped content, which contains the bits that
        the client-side on-load script trigger will execute the index.js
        defined for this mold; if this is not desired, simply call the
        render method instead.
        """

        template = self.load_mold(mold_id)

        kwargs = {}
        kwargs.update(data)
        kwargs['_nunja_data_'] = 'data-nunja="%s"' % mold_id
        kwargs['_template_'] = template
        kwargs['_wrapper_tag_'] = wrapper_tag

        return self._core_template_.render(**kwargs)

    def render(self, mold_id, data):
        """
        Render a mold `mold_id`.  No wrappers are applied as only the
        default template defined for the mold is rendered.
        """

        template = self.load_mold(mold_id)
        return template.render(**data)


class JinjaEngine(object):
    """
    Jinja only engine.

    This takes a jinja template registry for use of loading and
    rendering of templates.
    """

    def __init__(self, registry=JINJA_TEMPLATE_REGISTRY_NAME, env=None):
        """
        By default, the engine can be created without arguments which
        will initialize using the default registry.

        It is possible to initialize using other arguments, but this is
        unsupported by the main system, and only useful for certain
        specialized implementations.
        """

        self.registry = (
            registry
            if isinstance(registry, JinjaTemplateRegistry) else get(registry)
        )
        self.env = env if env else Environment(
            autoescape=True,
            loader=NunjaLoader(self.registry)
        )
        # this filter is to match with nunjucks version (which calls
        # JSON.stringify in JavaScript); construct a partial which is a
        # callable to json.dumps with default parameters that mimic the
        # JavaScript version of the called function.
        self.env.filters['dump'] = partial(
            json.dumps, sort_keys=True, separators=(',', ':'))

    def lookup_path(self, name):
        """
        Lookup the path of the underlying resource identified by name
        through the loader.
        """

        return self.registry.verify_path(name)

    def fetch_path(self, name):
        """
        Fetch contents from the path retrieved via lookup_path.

        No caching will be done.
        """

        with codecs.open(self.lookup_path(name), encoding='utf-8') as fd:
            return fd.read()

    def load_template(self, name):
        """
        Load the template identified by name as found in the registry.
        """

        return self.env.get_template(name)

    def render_template(self, name, data):
        """
        Render a template.
        """

        template = self.load_template(name)
        return template.render(**data)

File: src/nunja/engine.js
'use strict';

var nunjucks = require('nunjucks');
var registry = require('nunja/registry');
var loader = require('nunja/loader');
var utils = require('nunja/utils');

var $ = utils.$;

/* istanbul ignore next */
try {
    // The precompiled core is definitely required
    require('nunja/__core__');
}
catch(e) {
    console.log(
        'precompiled core templates missing; performance hit may result');
}

var _registry = new registry.Registry();

// Default environment that includes the nunja specific loader, and
// make use of autoescape as it's better to untrust templates by
// default.  To do unsafe actions one must include the safe filter,
// and this should stick out in auditing while keeping safety first.
var env = new nunjucks.Environment(new loader.NunjaLoader(_registry), {
    'autoescape': true,
});

// If any of these are overridden, we don't really support them but
// it's there if this later gets extended to do whatever other stuff
// that is needed.
var default_kwargs = {
    '_required_template_name': registry.REQ_TMPL_NAME,
    '_wrapper_name': registry.DEFAULT_WRAPPER_NAME,
    '_wrapper_tag_': registry.DEFAULT_WRAPPER_TAG,
    'env': env,
    'registry': _registry,
};

var Engine = function(kwargs) {
    for (var key in default_kwargs) {
        this[key] = kwargs[key] || default_kwargs[key];
    }

    // Load the default template.
    this['_core_template_'] = this.load_mold(this._wrapper_name);
};

// global scan function - looks up data-nunja ids and look up
// the default script, execute it on the afflicted element.
Engine.prototype.scan = function(context) {
    /*
    Collates and returns all elements with data-nunja attributes.
    */
    return $('[data-nunja]', context);
};

Engine.prototype.query_template = function (name) {
    /*
    Query for the template's loaded state.  Returns true if template can
    be loaded synchronously, false if otherwise.
    */

    // Two places that the template can be loaded: either through the
    // window.nunjucksPrecompiled or in the requirejs framework, if
    // available.

    return (
        (name in window.nunjucksPrecompiled) || (
            typeof requirejs !== "undefined" &&
            requirejs.defined('text!' + name)
        )
    );
};

Engine.prototype.load_template = function (name, cb) {
    /*
    Load the template identified by this name.
    */
    return this.env.getTemplate(name, cb);
};

Engine.prototype.load_mold = function (mold_id, cb) {
    /*
    Loads the default template for the mold identified by mold_id
    */
    return this.load_template(
        mold_id + '/' + this._required_template_name, cb);
};

Engine.prototype.load_element = function(element, cb) {
    /*
    Process this element.  This also ensures the template is
    available at some point.

    Returns the mold_id associated with this.
    */
    var mold_id = element.attributes.getNamedItem('data-nunja').value;
    this.load_mold(mold_id, cb);
    return mold_id;
};

Engine.prototype.init_element = function(element) {
    /*
    Triggers the main.init function associated with the mold of the
    element.  This should only be called once ever per element.

    element - The element we want.
    */
    var mold_id = this.load_element(element);
    var entry_point = mold_id + '/index';
    require([entry_point], function(main) {
        if (main && main.init instanceof Function) {
            main.init(element);
        }
    }, function() {
        // TODO whether have this be a definable behavior, i.e. whether
        // to leave undefine instead so import can be attempted again,
        // or that it be hard-defined to be nothing like so:
        // define(entry_point, [], {});
        // requirejs.undef(entry_point);
    });
};

Engine.prototype.do_onload = function (content) {
    /*
    The onLoad trigger.

    This scans for all relevant nodes that are related to nunja
    molds, and executes the relevant script for the relevant
    elements.
    */
    var elements = this.scan(content);
    var self = this;
    // elements.forEach(function (element, index, array) {
    elements.forEach(function (element) {
        self.init_element(element);
    });
};

Engine.prototype.execute = function (mold_id, data, cb) {
    /*
    Execute the complete template, which renders a parent block
    enclosing with the template that provides the nunja-data
    attribute.
    */
    var self = this;

    var _execute = function(error, template) {
        var params = data || {};
        params['_nunja_data_'] = 'data-nunja="' + mold_id + '"';
        params['_template_'] = template;
        params['_wrapper_tag_'] = self._wrapper_tag_;
        return self._core_template_.render(params, cb);
    };

    if (cb instanceof Function) {
        this.load_mold(mold_id, _execute);
    }
    else {
        var template = this.load_mold(mold_id);
        return _execute(null, template);
    }
};

Engine.prototype.populate = function (element, data, cb) {
    /*
    Populate the element that contains a data-nunja identifier with the
    data involved.  If a callback is provided, do so asynchronously.
    */
    var mold_id = element.getAttribute('data-nunja');

    if (!(cb instanceof Function)) {
        element.innerHTML = this.render(mold_id, data);
        this.init_element(element);
    }
    else {
        var self = this;
        this.render(mold_id, data, function(err, result) {
            // TODO handle err if there is an error somewhere...
            element.innerHTML = result;
            self.init_element(element);
            cb();
        });
    }

};

Engine.prototype.render = function (mold_id, data, cb) {
    /*
    Very simple, basic rendering method.

    The standard calling is done without the callback, which follow a
    synchronous flow.  If the callback ``cb`` is provided, the calling
    will be done asynchronously, where the cb must accept an error
    object produced by the template.render, and also the result of the
    rendering.
    */
    if (cb instanceof Function) {
        this.load_mold(mold_id, function(err, tmpl) {
            tmpl.render(data, cb);
        });
    }
    else {
        var template = this.load_mold(mold_id);
        var results = template.render(data);
        return results;
    }
};

exports.Engine = Engine;
